File: vim/custom_script/work_order/work_order.js
frappe.ui.form.on('Work Order', {
    setup : function(frm){
        frm.fields_dict['reusable_items'].grid.get_field('asset_name').get_query = function(frm, cdt, cdn) {
			var child = locals[cdt][cdn];
			return{
                query: "vim.custom_script.work_order.work_order.get_asset_list",
				filters: {
					'item_code': child.item_code
                    // ,
					// 'maximum_usage_count': [">", "used_count"]
				}
			}
		}
        frm.fields_dict['reusable_items'].grid.get_field('item_code').get_query = function(frm, cdt, cdn) {
			var child = locals[cdt][cdn];
			return{
				filters: {
					'reusable_item': 1
				}
			}
		}
    },
    //  refresh:function(frm){
    //     if (frm.doc.fg_warehouse == frm.doc.source_warehouse){
    //         var set_source = false;
    //         if(frm.doc.required_items)
    //         for(var i = 0; i<frm.doc.required_items.length;i++)
    //         {
    //             if( frm.doc.required_items[i].source_warehouse){
    //                 set_source = true;
    //                 break; 
    //             }
                
    //         }
    //         else 
    //         set_source = true;        
    //         if(!set_source){
    //             frm.set_value('source_warehouse', "");
    //             frm.set_value('source_warehouse', frm.doc.fg_warehouse);
    //         }
    //     }
    // },
    sales_order:function(frm){
    if(frm.doc.sales_order){
    frappe.db.get_value('Sales Order', frm.doc.sales_order, 'set_warehouse' ,function(r) {
        frm.set_value('fg_warehouse', r.set_warehouse);
        frm.set_value('source_warehouse', r.set_warehouse);
        frm.set_value('wip_warehouse', r.set_warehouse);

        frm.refresh_fields();
      });

    }
    },
    bom_no:function(frm){
        frm.set_value('reusable_items', []);
        frm.refresh_fields();

        if(frm.doc.bom_no){
            console.log(frm.doc.bom_no,"bom")
            frappe.call({
                "method": "vim.custom_script.work_order.work_order.get_item_list",
                args:{bom:frm.doc.bom_no},
                async: true,
                callback: function (r) {
            console.log(r,"reusable")

                    for(var i =0 ; i<r.message.length;i++)
                    {var childTable = cur_frm.add_child("reusable_items");
                    childTable.item_code=r.message[i];
                    cur_frm.refresh_fields("reusable_items");}
                }})
        frm.refresh_fields();
        }
        },
    before_save:function(frm){
        if(frm.doc.reusable_items)
        for(var i = 0; i<frm.doc.reusable_items.length;i++)
        {
            console.log(i,"i");
            if(frm.doc.reusable_items[i].balance_count <= 0){
                frappe.validated = false;
                frappe.throw("Item with no balance at row  "+frm.doc.reusable_items[i].idx)
            }
            if(!frm.doc.reusable_items[i].asset_name)
            frappe.call({
            "method": "vim.custom_script.work_order.work_order.check_asset_name",
            args:{item_code:frm.doc.reusable_items[i].item_code},
            async: false,
            callback: function (r) {
                // if (r.message["error"]){
                    if(r.message["asset"]){
                       
                        // {
                            frappe.validated = false;
                            frappe.throw("Asset Name is mandatory for row "+frm.doc.reusable_items[i].idx)}
                    // }
                    //":"Asset Name is mandatory for row {}".format(item.idx)
                   // :"Item with no balance at row {}".format(item.idx)
                // }
            }})}
    }
 });

 frappe.ui.form.on('Reusable Items', {
    asset_name:function(frm , cdt ,cdn){
	var child = locals[cdt][cdn];
    // console.log(child.asset_name,"child.asset_name")
    if(child.asset_name){
         frappe.db.get_value('Asset', child.asset_name, ['maximum_usage_count','used_count'],
        function(r) {
            // console.log(r)
            var asset_value =r
            var maximum_usage_count = asset_value["maximum_usage_count"];
            var used_count = asset_value["used_count"];
            // console.log(asset_value,asset_value.message,maximum_usage_count,used_count)
            child.maximum_usage_count = maximum_usage_count;
            child.used_count = used_count;
            child.balance_count = maximum_usage_count-used_count;
            frm.refresh_fields();
          });
       
    }
    },
    item_code:function(frm , cdt ,cdn){
        var child = locals[cdt][cdn];
        // console.log(child.asset_name,"child.asset_name")
        if(child.item_code){
             frappe.db.get_value('Item', child.item_code, ['maximum_usage_count','used_count'],
            function(r) {
                // console.log(r)
                var asset_value =r
                var maximum_usage_count = asset_value["maximum_usage_count"];
                var used_count = asset_value["used_count"];
                // console.log(asset_value,asset_value.message,maximum_usage_count,used_count)
                child.maximum_usage_count = maximum_usage_count;
                child.used_count = used_count;
                child.balance_count = maximum_usage_count-used_count;
                frm.refresh_fields();
              });
           
        }
        }
 });


//  frappe.ui.form.on("Work Order", "validate", function(frm, cdt, cdn) { 
// 	$.each(frm.doc.reusable_items || [], function(i, d) {
//         var Asset  = frappe.get_doc("Asset",{"item_code":d.item_code})
//         console.log(Asset&&!d.asset_name);
//         (Asset && !d.asset_name)
// 	    {
// 			frappe.throw("Asset Name is mandatory for row "+d.idx);
// 		}
// 	})
// });

File: vim/custom_script/work_order/work_order.py
# -*- coding: utf-8 -*-
# Copyright (c) 2021, Avu and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from collections import OrderedDict
from frappe import _
from frappe.model.document import Document

# @frappe.whitelist()
def validate(self,method):
	# check_asset_name(self)
	set_reusable_items(self)

@frappe.whitelist()
def check_asset_name(item_code):
	if frappe.db.exists({"doctype":"Asset","item_code":item_code}) :
				return {"asset":True}
	else :
				return {"asset":False}
			

def set_reusable_items(self):
	if self.bom_no and self.reusable_items == [] :
		items = get_item_list(self.bom_no)
		maximum_usage_count = 0
		used_count =0 
		balance_count =0
		for item in items:
			if not frappe.db.exists({"doctype":"Asset","item_code":item}):
				item_doc = frappe.get_doc("Item",{"item_code":item})
				maximum_usage_count = item_doc.maximum_usage_count
				used_count = item_doc.used_count
				balance_count = maximum_usage_count-used_count
			self.append('reusable_items', {
				"item_code": item,
				"maximum_usage_count":maximum_usage_count,
				"used_count":used_count,
				"balance_count":balance_count
			})


@frappe.whitelist()
def get_item_list(bom):
	result_list = []
	bom_doc = frappe.get_doc("BOM",bom)
	for item in bom_doc.items:
		item = frappe.get_doc("Item",{"item_code":item.item_code})
		if item  and item.reusable_item:
			result_list.append(item.item_code)

	# unitdetail= frappe.db.sql(unitdetail, as_dict=True)	
	# for tkt in unitdetail:
	# 	unitdetail_data = OrderedDict()
	# 	unitdetail_data['name'] = tkt.get('name') or ''				
	# 	result_list.append(unitdetail_data)
	
	return result_list	

@frappe.whitelist()
def get_asset_list(doctype, txt, searchfield, start, page_len, filters):
	# frappe.errprint([filters,"""select name from `tabAsset` where maximum_usage_count > used_count and item_code = '{}'""".format(filters["item_code"])])
	return frappe.db.sql("""select name from `tabAsset` where maximum_usage_count > used_count and item_code ='{}'""".format(filters["item_code"]))
